[webkitapipy] Allowed selector names are ambiguous between classes, leading to false positives

emw-apple opened the pull request:
#### b9fa7bcf117064356a95aae272fe426d65f30610
<pre>
[webkitapipy] Allowed selector names are ambiguous between classes, leading to false positives
<a href="https://bugs.webkit.org/show_bug.cgi?id=296133">https://bugs.webkit.org/show_bug.cgi?id=296133</a>

Reviewed by Mike Wyrzykowski and Brianna Fan.

It&apos;s possible for a selector to be implemented multiple times in the
SDK, where some implementations are SPI methods and others are API. In
this situation, audit-spi would prevent us from listing that selector as
an allowed declaration; it knows of the API method and emits an
UnnecessaryAllowedName diagnostic.

But if we are indeed binding to an SPI method, we want to be able to
track it in the allowlist, even if the selector&apos;s access is ambiguous.
To support this, change the allowlist format to use a class-bound
selector name. Instead of a string selector:

    &quot;initWithQueue:&quot;

we&apos;ll now write:

    { name = &quot;initWithQueue:&quot;, class = &quot;CoreTelephonyClient&quot; }

to allow -[CoreTelephonyClient initWithQueue:] specifically. When the
class is unknown, such as when being pre-populated by the sellist, a
class of &quot;?&quot; represents any receiver.

To support this, have APIReport (binary analysis) and SDKDB (SDK
analysis) ingestions track the class that each known selector is
implemented on. Add a clause to the audit() query to only match an
allowlist entry with a class binding if the `class_name`s match.

Reformat the selectors in existing allowlists. Add a few new clauses
that work towards making the internal iOS 26 build pass audit-spi.

* Source/JavaScriptCore/Configurations/AllowedSPI-legacy.toml:
* Source/JavaScriptCore/Configurations/AllowedSPI.toml:
* Source/WebCore/Configurations/AllowedSPI-legacy.toml:
* Source/WebCore/Configurations/AllowedSPI.toml:
* Source/WebGPU/Configurations/AllowedSPI-legacy.toml:
* Source/WebKit/Configurations/AllowedSPI-legacy.toml: Clarify some
  existing allowed selectors to be bound to specific SPI methods. These
  were discovered during work to support SPI checking in iOS 26 -- the
  selectors are all SPI methods, but there are unrelated API methods
  with the same name.

* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKitLegacy/mac/Configurations/AllowedSPI-legacy.toml:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowedSPI.Selector):
(AllowList.from_dict): As a drive-by enhancement, fail when an allowlist entry has
  an unrecognized key. This helps catch typos instead of silently
  ignoring them.
(AllowList.from_file): As a drive-by enhancement, report the file being
  parsed when handling a decode error.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
(APIReport):
(APIReport.Selector):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB._initialize_db):
(SDKDB.add_partial_sdkdb):
(SDKDB):
(SDKDB._add_partial_sdkdb):
(SDKDB.InsertionKind.statement):
(SDKDB._add_api_report):
(SDKDB._add_allowlist):
(SDKDB.add_for_auditing):
(SDKDB.audit):
(SDKDB._add_symbol):
(SDKDB._add_objc_class):
(SDKDB._add_objc_selector):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py: In
  addition to some new test cases, refactor the allowlist fixture logic
  to make it easier to test with different allowlists. Add cases
  which ingest SDKDBs.
(TestSDKDB.add_partial_sdkdb):
(TestSDKDB):
(TestSDKDB.add_allowlist):
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):
(TestSDKDB.test_audit_unnecessary_allow_from_selector):
(TestSDKDB.test_audit_allowed_fully_qualified_selector):
(TestSDKDB.test_audit_api_in_loaded_and_unloaded_library):

Canonical link: <a href="https://commits.webkit.org/298571@main">https://commits.webkit.org/298571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20407b63f2fce720f3a766062b0b7364b44f1cdd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115586 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25772 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121633 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66170 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0a35dd4f-7490-4392-8ef1-294d99594613) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117475 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35930 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87797 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42490 "An unexpected error occured. Recent messages:Printed configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7678331e-d2fc-4801-be62-d97beed1abd6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118534 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103734 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68195 "Passed tests") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/114960 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21849 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65312 "Built successfully") | | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107720 "Found 1 new JSC stress test failure: stress/catch-from-function-inlining-simd-function.js.ftl-eager-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98034 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21971 "Found 1 new test failure: fast/attachment/mac/wide-attachment-image-controls-basic.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124785 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114112 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42481 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31842 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96558 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99924 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96344 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24573 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41596 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19453 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38356 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47944 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41849 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45181 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43564 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c126edec4ec474544fc801baa62d91856a567501. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48192

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/144daa0c6da5e6a808f99c0901fd205a46eb1698. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48192

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/20407b63f2fce720f3a766062b0b7364b44f1cdd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48192

briannafan commented:
I just remembered, do we have to update `Tools/Scripts/libraries/webkitapipy/webkitapipy/reporter.py` with the new selector format? IIRC, the template is in the old string selector format.

emw-apple committed:
[webkitapipy] Allowed selector names are ambiguous between classes, leading to false positives
https://bugs.webkit.org/show_bug.cgi?id=296133

Reviewed by Mike Wyrzykowski and Brianna Fan.

It's possible for a selector to be implemented multiple times in the
SDK, where some implementations are SPI methods and others are API. In
this situation, audit-spi would prevent us from listing that selector as
an allowed declaration; it knows of the API method and emits an
UnnecessaryAllowedName diagnostic.

But if we are indeed binding to an SPI method, we want to be able to
track it in the allowlist, even if the selector's access is ambiguous.
To support this, change the allowlist format to use a class-bound
selector name. Instead of a string selector:

    "initWithQueue:"

we'll now write:

    { name = "initWithQueue:", class = "CoreTelephonyClient" }

to allow -[CoreTelephonyClient initWithQueue:] specifically. When the
class is unknown, such as when being pre-populated by the sellist, a
class of "?" represents any receiver.

To support this, have APIReport (binary analysis) and SDKDB (SDK
analysis) ingestions track the class that each known selector is
implemented on. Add a clause to the audit() query to only match an
allowlist entry with a class binding if the `class_name`s match.

Reformat the selectors in existing allowlists. Add a few new clauses
that work towards making the internal iOS 26 build pass audit-spi.

* Source/JavaScriptCore/Configurations/AllowedSPI-legacy.toml:
* Source/JavaScriptCore/Configurations/AllowedSPI.toml:
* Source/WebCore/Configurations/AllowedSPI-legacy.toml:
* Source/WebCore/Configurations/AllowedSPI.toml:
* Source/WebGPU/Configurations/AllowedSPI-legacy.toml:
* Source/WebKit/Configurations/AllowedSPI-legacy.toml: Clarify some
  existing allowed selectors to be bound to specific SPI methods. These
  were discovered during work to support SPI checking in iOS 26 -- the
  selectors are all SPI methods, but there are unrelated API methods
  with the same name.

* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKitLegacy/mac/Configurations/AllowedSPI-legacy.toml:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowedSPI.Selector):
(AllowList.from_dict): As a drive-by enhancement, fail when an allowlist entry has
  an unrecognized key. This helps catch typos instead of silently
  ignoring them.
(AllowList.from_file): As a drive-by enhancement, report the file being
  parsed when handling a decode error.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
(APIReport):
(APIReport.Selector):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB._initialize_db):
(SDKDB.add_partial_sdkdb):
(SDKDB):
(SDKDB._add_partial_sdkdb):
(SDKDB.InsertionKind.statement):
(SDKDB._add_api_report):
(SDKDB._add_allowlist):
(SDKDB.add_for_auditing):
(SDKDB.audit):
(SDKDB._add_symbol):
(SDKDB._add_objc_class):
(SDKDB._add_objc_selector):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py: In
  addition to some new test cases, refactor the allowlist fixture logic
  to make it easier to test with different allowlists. Add cases
  which ingest SDKDBs.
(TestSDKDB.add_partial_sdkdb):
(TestSDKDB):
(TestSDKDB.add_allowlist):
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):
(TestSDKDB.test_audit_unnecessary_allow_from_selector):
(TestSDKDB.test_audit_allowed_fully_qualified_selector):
(TestSDKDB.test_audit_api_in_loaded_and_unloaded_library):

Canonical link: https://commits.webkit.org/298571@main
